server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorDriver.java

```diff
@@ -456,63 +518,93 @@ public void moveSegmentOut(final String sequenceName, final List<SegmentIdentifi
    * @param committer committer
    *
    * @return a {@link ListenableFuture} for the publish task which removes published {@code sequenceNames} from
-   *         {@code activeSegments} and {@code publishPendingSegments}
+   * {@code activeSegments} and {@code publishPendingSegments}
    */
   public ListenableFuture<SegmentsAndMetadata> publishAll(
       final TransactionalSegmentPublisher publisher,
       final Committer committer
   )
   {
-    final List<SegmentIdentifier> theSegments;
-    synchronized (activeSegments) {
-      final List<String> sequenceNames = ImmutableList.copyOf(publishPendingSegments.keySet());
-      theSegments = sequenceNames.stream()
-                                 .map(publishPendingSegments::remove)
-                                 .filter(Objects::nonNull)
-                                 .flatMap(Collection::stream)
-                                 .collect(Collectors.toList());
-      sequenceNames.forEach(activeSegments::remove);
+    final List<String> theSequences;
+    synchronized (segments) {
+      theSequences = ImmutableList.copyOf(segments.keySet());
     }
-    return publish(publisher, wrapCommitter(committer), theSegments);
+    return publish(publisher, wrapCommitter(committer), theSequences);
   }
 
   /**
    * Execute a task in background to publish all segments corresponding to the given sequence names.  The task
    * internally pushes the segments to the deep storage first, and then publishes the metadata to the metadata storage.
    *
-   * @param publisher segment publisher
-   * @param committer committer
+   * @param publisher     segment publisher
+   * @param committer     committer
    * @param sequenceNames a collection of sequence names to be published
    *
    * @return a {@link ListenableFuture} for the submitted task which removes published {@code sequenceNames} from
-   *         {@code activeSegments} and {@code publishPendingSegments}
+   * {@code activeSegments} and {@code publishPendingSegments}
    */
   public ListenableFuture<SegmentsAndMetadata> publish(
       final TransactionalSegmentPublisher publisher,
       final Committer committer,
       final Collection<String> sequenceNames
   )
   {
-    final List<SegmentIdentifier> theSegments;
-    synchronized (activeSegments) {
-      theSegments = sequenceNames.stream()
-                                 .map(publishPendingSegments::remove)
-                                 .filter(Objects::nonNull)
-                                 .flatMap(Collection::stream)
-                                 .collect(Collectors.toList());
-      sequenceNames.forEach(activeSegments::remove);
+    final List<SegmentIdentifier> theSegments = new ArrayList<>();
+    synchronized (segments) {
+      sequenceNames.stream()
+                   .filter(sequenceName -> !publishingSequences.contains(sequenceName))
```

Comment:
The motivation behind using this in-memory Set is to prevent duplicate publishes of same sequence during the lifetime of this driver. Just the persisted segment state cannot be used for de-duplication because upon task restart some segments may be in `PUBLISHING` state because they were being published before restart but could not finish. So, the call to publish them again should actually start publishing them again as it is not a duplicate request in the current lifetime of driver.

The thing I am missing here though is a statement like `publishingSequences.add(sequenceName)` in `forEach`. I will add that.

Reply:
@pjain1 I'm working on fixing a bug in indexTask and this needs to refactor AppenderatorDriver. To avoid making another bug, I want to make sure what this code means.

> upon task restart some segments may be in PUBLISHING state because they were being published before restart but could not finish. So, the call to publish them again should actually start publishing them again as it is not a duplicate request in the current lifetime of driver.
- The segment state is changed to ```PUBLISHING```, but is never checked. Why is it needed?
- We can assume that all segments in the ```segments``` map are not published yet. If this is acceptable, is it possible to achieve the same thing by republishing segments without checking their state when publish() is called? Or, is it possible that publish() is called for the same sequence multiple times?

Reply:
At any point of time, segment can be in 3 state -
 - ACTIVE - Data is being added actively to it.
- INACTIVE - No data will be added to this segment, call to `moveSegmentOut` causes the segment to get into this state. Note that this segment is not yet started publishing by the AppenederatorDriver.
- PUBLISHING - Segment is being actively published, the segment information will be removed from `segments` map once publish succeeds.

Currently, the driver does not distinguish between INACTIVE and PUBLISHING state for making decision like preventing duplicate publishes. The reason is that driver's public publishing methods does not enable the caller to publish at segment level, it allows publishing at sequence level. So, `publishingSequences` Set is used to prevent duplicate sequence publishes. This is my understanding of the code and on the basis of this here are the answers to your questions -

> The segment state is changed to PUBLISHING, but is never checked. Why is it needed?

It looks like `PUBLISHING` state in not required and just changing the state to `INACTIVE` should work unless some task actually want to distinguish between moved out segments and segments that are being published.

>We can assume that all segments in the segments map are not published yet. If this is acceptable, is it possible to achieve the same thing by republishing segments without checking their state when publish() is called? Or, is it possible that publish() is called for the same sequence multiple times?

Currently, segments in the `segments` map might be in either active, inactive or being published state. The segment information is removed only when it is successfully published using a `Future` callback in `publish` method.

Feel free to ask for any more information it its still not clear.

Reply:
@pjain1 thanks for the detailed explanation. I have one more follow-up question. 

> publishingSequences Set is used to prevent duplicate sequence publishes.

When is it possible to publish the same sequence multiple times? Here are my understandings.

- A new sequence is created whenever the supervisor sets a new end offset.
- The sequence name is always increased by 1 when a new sequence is created, so it's unique for each sequence and cannot be duplicated.
- Once a sequence is published, it is removed from the `segments` map.
- Upon task restart, the restored segments in the `segments` map are not published yet. In other words, the segments which are not in the `segments` map are already published.

Reply:
Duplicate publishes can only happen when a caller calls `publish` for some sequence multiple times or calls `publishAll` multiple times or some combination of both. I believe Kafka Index task does not do that so it should be ok. However, I see that IndexTask calls `publishAll` twice at lines 711 and 743, so it needs to be checked if `publishingSequences ` Set is removed then no problem will happen and also in future if these methods are used anywhere else nothing problematic happens.

Reply:
Thanks! IndexTask actually calls publish() synchronously, so it should also be fine.

Reply:
If you are planning to remove `publishingSequences` Se then I just noticed that calling `publishAll` twice one after another immediately can be problematic as the code is relying on Future callback to remove the published sequences from `segments` map. However, as far as I know callbacks on future are not guaranteed to be executed before `get` on that future returns. So, it might happen that the sequence keys are not removed from the map before second call to `publishAll` happens so please check this edge case.

Also I think while removing sequence keys from map in the callback using this statement -
 `sequenceNames.forEach(segments::remove);` should be done is a `synchronized (segments)` block. I should have done that, its my mistake. Does this makes sense ?

Reply:
Yeah, it makes sense. I'll fix them too. Thanks for sharing!